runtime/tests/checker/interface_test.go

```diff
@@ -1363,52 +1363,74 @@ func TestCheckInterfaceWithFieldHavingStructType(t *testing.T) {
 
 	t.Parallel()
 
-	for _, firstKind := range common.CompositeKindsWithFieldsAndFunctions {
-		for _, secondKind := range common.CompositeKindsWithFieldsAndFunctions {
+	test := func(firstKind, secondKind common.CompositeKind) {
 
-			testName := fmt.Sprintf(
-				"%s/%s",
-				firstKind.Keyword(),
-				secondKind.Keyword(),
-			)
+		testName := fmt.Sprintf(
+			"%s in %s",
+			firstKind.Keyword(),
+			secondKind.Keyword(),
+		)
 
-			t.Run(testName, func(t *testing.T) {
+		t.Run(testName, func(t *testing.T) {
 
-				_, err := ParseAndCheck(t,
-					fmt.Sprintf(
-						`
+			t.Parallel()
+
+			_, err := ParseAndCheck(t,
+				fmt.Sprintf(
+					`
                           %[1]s S {}
 
                           %[2]s interface I {
                               s: %[3]sS
                           }
                         `,
-						firstKind.Keyword(),
-						secondKind.Keyword(),
-						firstKind.Annotation(),
-					),
-				)
+					firstKind.Keyword(),
+					secondKind.Keyword(),
+					firstKind.Annotation(),
+				),
+			)
 
-				// `firstKind` is the nested composite kind.
-				// `secondKind` is the container composite kind.
-				// Resource composites can only be nested in resource composite kinds.
+			// `firstKind` is the nested composite kind.
+			// `secondKind` is the container composite kind.
+			// Resource composites can only be nested in resource composite kinds.
 
-				if firstKind == common.CompositeKindResource {
-					switch secondKind {
-					case common.CompositeKindResource,
-						common.CompositeKindContract:
+			switch firstKind {
+			case common.CompositeKindResource:
+				switch secondKind {
+				case common.CompositeKindResource,
+					common.CompositeKindContract:
 
-						require.NoError(t, err)
+					require.NoError(t, err)
 
-					default:
-						errs := ExpectCheckerErrors(t, err, 1)
+				default:
+					errs := ExpectCheckerErrors(t, err, 1)
+
+					assert.IsType(t, &sema.InvalidResourceFieldError{}, errs[0])
+				}
+
+			case common.CompositeKindContract:
+				switch secondKind {
+				case common.CompositeKindResource,
+					common.CompositeKindStructure:
```

Comment:
This is saying that a contract can be a field in a resource or structure? That doesn't make sense to me -- I thought contracts were always the top-level unit of code.

Reply:
Contracts are just objects, just like structs, and resources. 
They can be stored in variables, passed to functions, and also put into a struct or resource as a field.

Now, this may seem odd, but could be useful when using interfaces.
For example, a function could return a result object which has a field that has a contract interface type, indicating to the caller what contract should be called next.

Note that this has nothing to do with declaration nesting: contracts can still not be declared inside resources or structs, this is purely about fields inside containers, i.e. your statement " "contracts [are] always the top-level unit of code" is still true.